src/wp-admin/includes/image.php

```diff
@@ -393,79 +439,114 @@ function _wp_make_subsizes( $new_sizes, $file, $image_meta, $attachment_id ) {
 
-	$editor = wp_get_image_editor( $file );
+		// Check if any of the new sizes already exist for this mime type.
+		if ( isset( $image_meta['sizes'] ) && is_array( $image_meta['sizes'] ) ) {
+			foreach ( $image_meta['sizes'] as $size_name => $size_meta ) {
+				if ( ! isset( $size_meta['sources'] ) ) {
+					$size_meta['sources'] = array();
+				}
 
-	if ( is_wp_error( $editor ) ) {
-		// The image cannot be edited.
-		return $image_meta;
-	}
+				if ( ! isset( $new_sizes['sources'][ $output_mime_type ] ) ) {
+					$new_sizes['sources'][ $output_mime_type ] = array();
+					continue;
+				}
 
-	// If stored EXIF data exists, rotate the source image before creating sub-sizes.
-	if ( ! empty( $image_meta['image_meta'] ) ) {
-		$rotated = $editor->maybe_exif_rotate();
+				/*
+				* Only checks "size name" so we don't override existing images even if the dimensions
+				* don't match the currently defined size with the same name.
+				* To change the behavior, unset changed/mismatched sizes in the `sizes` array in image meta.
+				*/
+				if ( array_key_exists( $size_name, $new_sizes['sources'][ $output_mime_type ] ) ) {
+					unset( $new_sizes[ $size_name ] );
+				}
+			}
+		} else {
+			$image_meta['sizes'] = array();
+		}
 
-		if ( is_wp_error( $rotated ) ) {
-			// TODO: Log errors.
+		if ( empty( $new_sizes ) ) {
+			// Nothing to do...
+			return $image_meta;
 		}
-	}
 
-	if ( method_exists( $editor, 'make_subsize' ) ) {
-		foreach ( $new_sizes as $new_size_name => $new_size_data ) {
-			$new_size_meta = $editor->make_subsize( $new_size_data );
+		/*
+		* Sort the image sub-sizes in order of priority when creating them.
+		* This ensures there is an appropriate sub-size the user can access immediately
+		* even when there was an error and not all sub-sizes were created.
+		*/
+		$priority = array(
+			'medium'       => null,
+			'large'        => null,
+			'thumbnail'    => null,
+			'medium_large' => null,
+		);
+
+		$new_sizes = array_filter( array_merge( $priority, $new_sizes ) );
 
-			if ( is_wp_error( $new_size_meta ) ) {
+		// If stored EXIF data exists, rotate the source image before creating sub-sizes.
+		if ( ! empty( $image_meta['image_meta'] ) ) {
+			$rotated = $editor->maybe_exif_rotate();
+
+			if ( is_wp_error( $rotated ) ) {
 				// TODO: Log errors.
-			} else {
-				// Save the size meta value.
-				$image_meta['sizes'][ $new_size_name ] = $new_size_meta;
-				wp_update_attachment_metadata( $attachment_id, $image_meta );
 			}
 		}
-	} else {
-		// Fall back to `$editor->multi_resize()`.
-		$created_sizes = $editor->multi_resize( $new_sizes );
 
-		if ( ! empty( $created_sizes ) ) {
-			$image_meta['sizes'] = array_merge( $image_meta['sizes'], $created_sizes );
-			wp_update_attachment_metadata( $attachment_id, $image_meta );
+		if ( method_exists( $editor, 'make_subsize' ) ) {
+			foreach ( $new_sizes as $new_size_name => $new_size_data ) {
+				$new_size_meta = $editor->make_subsize( $new_size_data, $output_mime_type );
```

Comment:
Since this method may be implemented by other custom `WP_Image_Editor` child classes, we need to definitely document this new parameter in the dev note. However, the current implementation could be more problematic than that, since those custom implementations wouldn't necessarily support the new parameter and therefore generate the same MIME type multiple times.

If nobody extends the `WP_Image_Editor` class with this method, we'd be fine. But the safer decision here would be to introduce an entirely separate method `set_mime_type`, and only run the extra MIME type generation logic _if_ the editor class used has support for that new method. That would mean it works for WP core, but it also would still fall back gracefully for custom implementations without causing problems.

Of course these custom implementations wouldn't actually get the new image format subsizes in that case, but they wouldn't get them with the current implementation either. So using a new method as a signal for the base class to tell WP core "I support multiple MIME types" is a safer solution.

Reply:
Couple of questions about that approach:
If we add `set_mime_type` I assume we would call that right before `make_subsize`? What about all the existing functions like `save` that already accept a mime_type parameter, would these use the set mime type when `set_mime_type` has been called? we would also need a way to unset the mime type I guess.

What about a new function `multi_resize_to_mime` (or something like that) that takes the mime_type parameter and otherwise is a copy of `multi_resize`, this would still prevent conflicts and maybe  make the change easier to support.

Reply:
current code still modifies `make_subsize` however it only supports multiple mimes when `make_subsize` is available; partially solves the issue raised. Can still add a new function to indicate support though, how about my proposed `multi_resize_to_mime`?